### ISSUE TITLE
Timeout in Redis XREAD: read tcp [...]: i/o timeout

### ISSUE THREAD

**Issue by julianbrost** [Icinga/icingadb#786]
We received a report of IcingaDB 1.2.0 crashing while logging the following:

```
Jun 21 18:58:45 icingadb[1628]: runtime-updates: Upserted 2 ServiceState items
Jun 21 18:58:57 icingadb[1628]: history-sync: Synced 2 state history items
Jun 21 18:59:05 icingadb[1628]: runtime-updates: Upserted 1 ServiceState items
Jun 21 18:59:17 icingadb[1628]: history-sync: Synced 1 state history items
Jun 21 18:59:25 icingadb[1628]: runtime-updates: Upserted 3 ServiceState items
Jun 21 18:59:37 icingadb[1628]: history-sync: Synced 3 state history items
Jun 21 18:59:45 icingadb[1628]: runtime-updates: Upserted 1 ServiceState items
Jun 21 19:00:45 icingadb[1628]: heartbeat: Lost Icinga heartbeat
Jun 21 19:01:45 icingadb[1628]: heartbeat: Waiting for Icinga heartbeat
Jun 21 19:02:54 icingadb[1628]: heartbeat: Waiting for Icinga heartbeat
Jun 21 19:02:54 icingadb[1628]: high-availability: Lost heartbeat
Jun 21 19:02:54 icingadb[1628]: Handing over
Jun 21 19:02:54 icingadb[1628]: heartbeat: Received Icinga heartbeat
Jun 21 19:02:54 icingadb[1628]: read tcp [::1]:50258->[::1]:6380: i/o timeout
                                can't perform "[xread block 1000 streams icinga:dump 1718887165045-4]"
                                github.com/icinga/icingadb/pkg/icingaredis.WrapCmdErr
                                        github.com/icinga/icingadb/pkg/icingaredis/utils.go:121
                                github.com/icinga/icingadb/pkg/icingaredis.(*Client).XReadUntilResult
                                        github.com/icinga/icingadb/pkg/icingaredis/client.go:204
                                github.com/icinga/icingadb/pkg/icingadb.(*DumpSignals).Listen
                                        github.com/icinga/icingadb/pkg/icingadb/dump_signals.go:64
                                main.run.func3.1
                                        github.com/icinga/icingadb/cmd/icingadb/main.go:188
                                golang.org/x/sync/errgroup.(*Group).Go.func1
                                        golang.org/x/sync@v0.7.0/errgroup/errgroup.go:78
                                runtime.goexit
                                        runtime/asm_amd64.s:1695
                                can't read dump signals
                                github.com/icinga/icingadb/pkg/icingadb.(*DumpSignals).Listen
                                        github.com/icinga/icingadb/pkg/icingadb/dump_signals.go:68
                                main.run.func3.1
                                        github.com/icinga/icingadb/cmd/icingadb/main.go:188
                                golang.org/x/sync/errgroup.(*Group).Go.func1
                                        golang.org/x/sync@v0.7.0/errgroup/errgroup.go:78
                                runtime.goexit
                                        runtime/asm_amd64.s:1695
```

refs https://github.com/Icinga/icinga-go-library/pull/23
ref/IP/54215

**Comment by Al2Klimov** [comment]
# My analysis from the ticket

(And my explanations why

* https://github.com/Icinga/icinga-go-library/pull/23

will fix the problem.)

## https://rt.icinga.com/Ticket/Display.html?id=54215#txn-906034

icingadb[1628]: read tcp [::1]:50258->[::1]:6380: i/o timeout

This can only be net.timeoutError, wrapped by a net.OpError with read, tcp. The stack is also traceable:

https://github.com/Icinga/icingadb/blob/v1.2.0/cmd/icingadb/main.go#L188
https://github.com/Icinga/icingadb/blob/v1.2.0/pkg/icingadb/dump_signals.go#L64
https://github.com/Icinga/icingadb/blob/v1.2.0/pkg/icingaredis/client.go#L197
https://github.com/redis/go-redis/blob/v9.5.1/stream_commands.go#L168
https://github.com/redis/go-redis/blob/v9.5.1/redis.go#L627
https://github.com/redis/go-redis/blob/v9.5.1/redis.go#L647
https://github.com/redis/go-redis/blob/v9.5.1/redis.go#L675
https://github.com/redis/go-redis/blob/v9.5.1/redis.go#L185
https://github.com/redis/go-redis/blob/v9.5.1/redis.go#L650
https://github.com/redis/go-redis/blob/v9.5.1/redis.go#L400
https://github.com/redis/go-redis/blob/v9.5.1/redis.go#L426
https://github.com/redis/go-redis/blob/v9.5.1/internal/pool/conn.go#L74

The 3rd-last of these is actually re-tried, based on the info from the 2nd-last, i.e. eff. this function:
https://github.com/redis/go-redis/blob/v9.5.1/error.go#L40-L73

Since net.OpError and net.timeoutError now return true via interface{Timeout()bool}, redis.shouldRetry() should also do this - if instructed to do so by its 2nd parameter retryTimeout bool. But the variable for this prevents this:
https://github.com/redis/go-redis/blob/v9.5.1/redis.go#L430

XREAD has a timeout - the same 1000ms that we want to BLOCK:
https://github.com/redis/go-redis/blob/v9.5.1/stream_commands.go#L165

I suggest not only retrying redis.Nil in icingaredis.Client#XReadUntilResult(), but also interface{Timeout()bool}. Or anything else that we usually re-trien. Alternatively, I would consider BLOCK -1, but https://github.com/redis/go-redis/issues/2131 stands in the way.

Translated with DeepL.com (free version)

## https://rt.icinga.com/Ticket/Display.html?id=54215#txn-906203

A BLOCK actually causes the XREAD to set a timeout on the socket:
https://github.com/redis/go-redis/blob/v9.5.1/internal/pool/conn.go#L70

But with a surcharge of 10s, otherwise it would also be too tight:
https://github.com/redis/go-redis/blob/v9.5.1/redis.go#L454

But what I actually wanted to say:

We set a BLOCK on our XREAD. https://github.com/Icinga/icingadb/blob/v1.2.0/pkg/icingaredis/client.go#L194
This tells the library to give up after this time.
And if a timeout occurs, it does exactly that and does not re-try. https://github.com/redis/go-redis/blob/v9.5.1/redis.go#L430
But at least it is honest and passes the timeout error.

(...)

And that's exactly why I would not only re-try redis.Nil, but also interface{Timeout()bool}&&Timeout().

Translated with DeepL.com (free version)

## https://rt.icinga.com/Ticket/Display.html?id=54215#txn-906274

> What I'm only now noticing: something was already strange in the Icinga DB process before
this error in the Icinga DB process, this delay between the two messages
between the two messages shouldn't actually be there:

Right, actually(TM) this message should come ASAP:
https://github.com/Icinga/icingadb/blob/v1.2.0/pkg/icingadb/ha.go#L243

At least the corresponding channel is triggered immediately:
https://github.com/Icinga/icingadb/blob/v1.2.0/pkg/icingaredis/heartbeat.go#L146-L147

However, it may well be that the channel receiver is still busy elsewhere with the previous heartbeat:
https://github.com/Icinga/icingadb/blob/v1.2.0/pkg/icingadb/ha.go#L221

This location even has its own deadline!
The previous heartbeat probably also triggered the reading of the dump singals.
All in all, on a completely crazy system, the only way to keep it from breaking (by timeout) is for it to normalize itself.

Translated with DeepL.com (free version)

## TL;DR

**The problem is a timeout. We are retrying timeouts in general. Just not here, yet.**

**Comment by julianbrost** [comment]
A read tcp i/o timeout during XREAD can actually triggered quite easily by simply sending the redis-server process a SIGSTOP, and a few seconds later:
```
2024-08-07T12:35:11.129Z	FATAL	icingadb	read tcp 172.18.0.36:59894->172.18.0.11:6379: i/o timeout
can't perform "[xread block 1000 streams icinga:schema 1722610884601-0]"
```

**Comment by tbauriedel** [comment]
ref/NC/824567
